metals/src/main/scala/scala/meta/internal/metals/callHierarchy/CallHierarchyItemBuilder.scala

```diff
@@ -0,0 +1,148 @@
+import scala.meta.internal.semanticdb.TextDocument
+import scala.meta.internal.{semanticdb => s}
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.CancelToken
+
+import org.eclipse.{lsp4j => l}
+
+private[callHierarchy] final class CallHierarchyItemBuilder(
+    workspace: AbsolutePath,
+    icons: Icons,
+    compilers: () => Compilers,
+    buildTargets: BuildTargets,
+) {
+  private def buildItem(
+      name: String,
+      kind: l.SymbolKind,
+      uri: String,
+      range: l.Range,
+      selectionRange: l.Range,
+      detail: String,
+      data: CallHierarchyItemInfo,
+  ): l.CallHierarchyItem = {
+    val item = new l.CallHierarchyItem(name, kind, uri, range, selectionRange)
+    item.setDetail(detail)
+    item.setData(data)
+    item
+  }
+
+  private def getClassContructorSymbol(
+      info: s.SymbolInformation
+  ): List[String] =
+    if (info.isClass)
+      info.signature
+        .asInstanceOf[s.ClassSignature]
+        .declarations
+        .flatMap(
+          _.symlinks.find(_.endsWith("`<init>`()."))
+        )
+        .toList ++ Option.when(info.isCase)(
+        info.symbol.replace("#", ".apply().")
+      ) // handling case class constructors
+    else Nil
+
+  /** Get the name of the class of a constructor */
+  private def getName(info: s.SymbolInformation): String =
+    if (info.isConstructor && info.displayName == "<init>")
+      info.symbol.slice(
+        info.symbol.lastIndexOf("/") + 1,
+        info.symbol.lastIndexOf("#"),
+      )
+    else info.displayName
+
+  private def getDetail(
+      signature: String,
+      visited: Array[String],
+      symbol: String,
+  ): String =
+    (if (visited.dropRight(1).contains(symbol)) icons.sync + " "
+     else "") + signature
+
+  private def getSignatureFromHover(hover: Option[l.Hover]): Option[String] =
+    (for {
+      hover <- hover
+      hoverContent <- hover.getContents().asScala.toOption
+      `match` <- """Symbol signature\*\*:\n```scala\n(.*)\n```""".r
+        .findFirstMatchIn(hoverContent.getValue)
+    } yield `match`.group(1))
+
+  private def buildItem(
+      symbolInformations: Seq[s.SymbolInformation],
+      source: AbsolutePath,
+      range: l.Range,
+      visited: Array[String],
+      symbol: String,
+      selectionRange: s.Range,
+      hover: Option[l.Hover],
+  ): Option[l.CallHierarchyItem] = {
+    symbolInformations
+      .find(_.symbol == symbol)
+      .map(info => {
+        buildItem(
+          getName(info),
+          info.kind.toLSP,
+          source.toURI.toString(),
+          range,
+          selectionRange.toLSP,
+          getDetail(
+            getSignatureFromHover(hover).getOrElse(""),
+            visited,
+            symbol,
+          ),
+          CallHierarchyItemInfo(
+            // When we search the call hierarchy of a class, we also need to search the symbol of the primary constructor
+            (Set(symbol) ++ getClassContructorSymbol(info)).toArray,
```

Comment:
This might be extracted to function (probably incorporated into `getClassContructorSymbol` + rename will be necessary).